src/group_theory/group_ring.lean

```diff
@@ -0,0 +1,216 @@
+
+variables {k : Type*} [semiring k] {G : Type*} [monoid G]
+
+lemma smul_of_eq_single (g : G) (r : k) :
+  r • (of k G g) = finsupp.single g r :=
+by simp only [mul_one, finsupp.smul_single', of_apply]
+
+lemma mk_linear_smul {P : Type*} {P' : Type*} [add_comm_monoid P] [add_comm_monoid P']
+  [module k P] [module k P']
+  [module (group_algebra k G) P] [module (group_algebra k G) P']
+  [is_scalar_tower k (group_algebra k G) P] [is_scalar_tower k (group_algebra k G) P']
+  (f : P →ₗ[k] P') (H : ∀ g x, f (of k G g • x) = of k G g • f x)
+  (r : group_algebra k G) (x : P) :
+  f (r • x) = r • f x :=
+begin
+  refine r.induction_on (λ g, by exact H g x) _ _,
+  { intros a b ha hb,
+    simp only [add_smul, f.map_add, ha, hb] },
+  { intros r a ha,
+    simp only [smul_assoc, f.map_smul, ha] }
+end
+
+/-- Makes a `k[G]`-linear map from a `k`-linear map that is also `G`-linear. -/
+@[simps] def mk_linear {P : Type*} {P' : Type*} [add_comm_monoid P] [add_comm_monoid P']
+  [module k P] [module k P']
+  [module (group_algebra k G) P] [module (group_algebra k G) P']
+  [is_scalar_tower k (group_algebra k G) P] [is_scalar_tower k (group_algebra k G) P']
+  (f : P →ₗ[k] P') (H : ∀ g x, f (of k G g • x) = of k G g • f x) :
+  P →ₗ[group_algebra k G] P' :=
+{ map_smul' := mk_linear_smul f H, ..f }
+
+/-- Makes a `k[G]`-linear isomorphism from a `k`-linear isomorphism which is also `G`-linear. -/
+@[simps] def mk_equiv {P : Type*} {P' : Type*} [add_comm_monoid P] [add_comm_monoid P']
+  [module k P] [module k P']
+  [module (group_algebra k G) P] [module (group_algebra k G) P']
+  [is_scalar_tower k (group_algebra k G) P] [is_scalar_tower k (group_algebra k G) P']
+  (f : P ≃ₗ[k] P') (H : ∀ g x, f (of k G g • x) = of k G g • f x) :
+  P ≃ₗ[group_algebra k G] P' :=
+{ ..f.to_equiv, ..mk_linear f.to_linear_map H }
+
+end
+
+variables {k : Type*} [comm_semiring k] {G : Type*} [group G] {n : ℕ}
+
+instance {H : Type*} [mul_action G H] : distrib_mul_action G (group_algebra k H) :=
+finsupp.comap_distrib_mul_action
+
+/-- If `k[H]` is a `k[G]`-module, then a `k`-linear map `f: k[H] →ₗ P` of `k[G]`-modules which
+satisfies `f(g • h) = g • f(h)` for all `g ∈ G, h ∈ H` is `k[G]`-linear. -/
+lemma map_smul_of_map_smul_of {H : Type*} [monoid H]
+  [module (group_algebra k G) (group_algebra k H)] {P : Type*} [add_comm_monoid P] [module k P]
+  [module (group_algebra k G) P] [is_scalar_tower k (group_algebra k G) (group_algebra k H)]
+  [is_scalar_tower k (group_algebra k G) P] (f : group_algebra k H →ₗ[k] P)
+  (h : ∀ (g : G) (x : H), f (of k G g • of k H x) = of k G g • f (of k H x))
+  (g : group_algebra k G) (x : group_algebra k H) : f (g • x) = g • f x :=
+begin
+  rw mk_linear_smul,
+  intros a b,
+  refine b.induction_on (by exact h a) _ _,
+  { intros s t hs ht,
+    simp only [smul_add, f.map_add, hs, ht] },
+  { intros r s hs,
+    simp only [smul_algebra_smul_comm, f.map_smul, hs] }
+end
+
+instance : smul_comm_class k G (group_algebra k (fin n → G)) :=
```

Comment:
This generalizes to
```suggestion
instance {M} [monoid M] [mul_action G M] : smul_comm_class k G (group_algebra k M)
```

Reply:
Thanks. I haven't generalised the subsequent two instances of this style because it messes up my later material; I hope that's okay, but if not, I can try and work something out